[RPC] Simplified bumpfee command.

mrbandrews opened the pull request:
This purports to be a simplified "bumpfee" RPC command.

This pull was motivated by the discussion in #7865, where at the end, the idea was floated of submitting a simplified version of the code. I put a little thought into how a simple "bumpfee" command should work and here is what I came up with:

1) The user should specify which output to decrement in order to pay increased fee. This saves us from trying to figure out which address(es) are "change" addresses. This is preferable not only because it simplifies the code, but because as far as I can tell, the code that identifies change outputs is not particularly robust and may be modified in the future. If it's desirable to have user-friendly code that figures out which output to decrement, perhaps that logic could be placed in the GUI?

2) The command should not add new inputs. In other words, if the user can't identify an output that can be decremented in an amount sufficient to bump the fee then the command should fail. It seems likely that the vast majority of the time, identifying such an output would not be a problem. Adding new inputs complicates the code because the size of the transaction increases, perhaps significantly, so then the user would have to pay more fee for that reason as well, as opposed to just bumping the fee on the tx as it currently exists.

3) If the tx has descendant transactions, the bumped tx has to pay the fees for all of the descendants since they will get evicted from the mempool, and the rule as I understand it is that replacing a tx cannot cause the total fees in the mempool to go down.

So, the required inputs are the txid and the output index.

The optional inputs are:

```
confirmTarget - perhaps the reason the user feels the need to bump the fee is because the original confirmTarget was too slow
totalFee - for full control, the user can simply specify how many satoshis to pay as a fee on the bumped transaction (this may be useful when the parent tx has children tx'es)
```

This pull includes a python test.

jonasschnelli commented:
Thanks! Looks good.
Concept ACK will review and test soon.

NicolasDorier commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2554,6 +2556,230 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+
+    // calculate the old fee and fee-rate
+    CAmount nDebit = wtx.GetDebit(ISMINE_SPENDABLE);
+    CAmount nOldFee = -(wtx.IsFromMe(ISMINE_SPENDABLE) ? wtx.GetValueOut() - nDebit : 0);
```
nit: Double negation makes it hards to read. I suggest to replace by

```
 CAmount nOldFee = wtx.IsFromMe(ISMINE_SPENDABLE) ? nDebit - wtx.GetValueOut() : 0;
```

mrbandrews replied:
Fixed.

NicolasDorier commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2554,6 +2556,230 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    CFeeRate minMempoolFeeRate = mempool.GetMinFee(GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000);
+    if (nNewFeeRate.GetFeePerK() < minMempoolFeeRate.GetFeePerK()) {
+        strError = strprintf("New fee rate (%s) is too low to get into the mempool (min rate: %s)", FormatMoney(nNewFeeRate.GetFeePerK()), FormatMoney(minMempoolFeeRate.GetFeePerK()));
+        throw JSONRPCError(RPC_MISC_ERROR, strError);
```
I'm not sure about this. The problem is that such case will rarely happen so applications risk to break during sudden peak usage, amplifying bad news. 

I suggest to silently bump fees to the minMempoolFeeRate.

NicolasDorier commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2554,6 +2556,230 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+            // and run the command again.
+            //
+            CAmount nFeesWithDescendants = it->GetModFeesWithDescendants();
+            if (nNewFee < nFeesWithDescendants + ::minRelayTxFee.GetFee(txSize)) {
```
nit: 
```
 nFeesWithDescendants + ::minRelayTxFee.GetFee(txSize)
```
Is repeated.

mrbandrews replied:
Fixed.

NicolasDorier commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2554,6 +2556,230 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    // Output must be able to pay the increased fee, without being reduced to dust
+    CMutableTransaction tx(wtx);
+    CTxOut* poutput = &(tx.vout[nOutput]);
+    if (poutput->nValue >= nDelta + poutput->GetDustThreshold(::minRelayTxFee)) {
```
additional check should be done:

```
nDelta >= poutput->GetDustThreshold(::minRelayTxFee);
```
Or it won't get propagated.

mrbandrews replied:
I checked earlier that the bumped fee pays for the new relay fee;  I'm not following as to the relationship between nDelta and dust.

luke-jr replied:
@NicolasDorier I also don't see a need for that...

NicolasDorier commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2554,6 +2556,230 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    // sign the new tx
+    CTransaction txNewConst(tx);
+    int nIn = 0;
+    for (auto it(tx.vin.begin()); it != tx.vin.end(); ++it, nIn++) {
```
can you use C++ loop style ?

mrbandrews replied:
Good idea, fixed.

NicolasDorier commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2554,6 +2556,230 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+        if (mi != pwalletMain->mapWallet.end() && (nIn < (int)(*mi).second.vout.size())) {
+            const CScript& scriptPubKey = (*mi).second.vout[(*it).prevout.n].scriptPubKey;
+            SignatureData sigdata;
+            if (!ProduceSignature(TransactionSignatureCreator(pwalletMain, &txNewConst, nIn, SIGHASH_ALL), scriptPubKey, sigdata))
```
You should use the same SIGHASH than the previous one.

mrbandrews replied:
Let's punt to separate PR.

NicolasDorier commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2554,6 +2556,230 @@ UniValue fundrawtransaction(const UniValue& params, bool fHelp)
+    int nIn = 0;
+    for (auto it(tx.vin.begin()); it != tx.vin.end(); ++it, nIn++) {
+        std::map<uint256, CWalletTx>::const_iterator mi = pwalletMain->mapWallet.find((*it).prevout.hash);
+        if (mi != pwalletMain->mapWallet.end() && (nIn < (int)(*mi).second.vout.size())) {
```
>  (nIn < (int)(*mi).second.vout.size())

I think this is a bug, why does the index of the input to sign should be below the number of vout of the parent ? Does not make sense.

mrbandrews replied:
Good catch, edited to bounds-check the prevout.

jonasschnelli commented:
Needs rebase and reviewers... setting 0.14 milestone.

mrbandrews commented:
Rebased and addressed feedback.

jonasschnelli commented:
Needs rebase again...

mrbandrews commented:
Rebased and edited to use JSONRPCRequest, consistent with #8788.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,234 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+
+    if (request.fHelp || request.params.size() < 2 || request.params.size() > 3)
+        throw runtime_error(
+                            "bumpfee \"txid\"\n"
```
This line is outdated

RHavar commented:
I think it's very important for this command figures out change output.  I understand that it's a bit messier and more fragile doing it here -- but it actually has enough information to do this, and by avoiding it, it just pushes that mess into the caller which comes at a very significant usability issue.


Compare:
Transaction stuck? Type `bumpfee $txid`

vs

Transaction stuck? First type `gettransaction $txid`, now copy the data from the "hex" field into `decoderawtransaction $hex`  now go through the "vout" fields until you find "addresses" nested in the "scriptPubKey". Now try figure out which is a change address. (I can't even see this exposed over rpc at all?). So maybe go `validateaddress $address` and check the `ismine` field. If one of them is yours, but the other one isn't -- then the one that is yours is a change address, so now go back through the results of `decoderawtransaction` to figure out which index that is. And finally `bumpfee $txid $index`.  Oh yeah, becareful to handle all the edge cases, like the transaction only having 1 output, some of the outputs not having address, both having the same address etc.


I think you'll find the usability problem of needing to know the index will prevent the majority of use by users and services for this =)

RHavar commented:
Also ideally, I think `txid` argument should be an array of transaction ids to bump. And then it creates a single transaction that bumps the fees on all of those. However, I imagine this can be done separately and lately as the `txid` argument could be overloaded to either accept a string or array of strings?

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,234 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        throw JSONRPCError(RPC_MISC_ERROR, strError);
+    }
+
+    CAmount nDelta = nNewFee - nOldFee;
```
There is kind of a weird deal here where nDelta can be <= 0 if totalFee is set, and will be accepted on a race condition with mempool, I would explicitly guard against this case by restricting nDelta > 0.

mrbandrews replied:
Done.

morcos replied:
@JeremyRubin how could nDelta be negative here?

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,234 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    CFeeRate nOldFeeRate(nOldFee, txSize);
+
+    // use the user-defined payTxFee if possible, otherwise use smartfee / fallbackfee
+    CAmount nNewFee = 0;
```
Using a nNewFee = 0 is fine here, it just slightly bothers me to use this as a null value when indeed 0 is a valid fee amount.

-1 would be an invalid fee amount (creates coins) so I have a slight preference to use this.

JeremyRubin commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,234 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            // and run the command again.
+            //
+            CAmount nFeesWithDescendantsPlusRelay = it->GetModFeesWithDescendants() + ::minRelayTxFee.GetFee(txSize);
+            if (nNewFee < nFeesWithDescendantsPlusRelay) {
```
I would suggest that this should also return in the error message the txid of the furthest child, and suggest bumping the fee on that one if possible to take advantage of ancestor fee based mining and keep txs valid?

JeremyRubin replied:
There is something kinda funky when a user is running in say blocksonlymode and doesn't know about any child transactions that may exist and therefore has trouble setting the fee correctly for those that they will invalidate.

This is probably a hard problem to solve; so I'm just pointing it out.

JeremyRubin reviewed (commented):
Overall seems like the right behavior for bumpfee, just a few minor suggestions.

It seems that the descendants fee adds additional complexity that probably can't be addressed in the scope of this PR but at least could merit some documentation so that a stuck user can find out what's wrong.

mrbandrews commented:
Addressed JeremyRubin feedback, edited the python test, and made a few other small changes I noticed with further testing. 

RHavar:  I understand your point but I still think it's better for this command to be low-level and not fragile.  A more user-friendly RPC (e.g., "bumpfeeauto" or something) could be layered on top, identifying the change output and then using this code.  Then if the change-output-identifying code breaks, it might break the user-friendly version but it wouldn't dismantle RBF entirely.

Victorsueca commented:
ACK
Tested on Windows x64
```
bumpfee 8c56b13830405a55ec4bc58b26b531f1b187d2349ee19bd0dd01aa835972929a 1

{
  "txid": "67d6af1a3e29e246eaef0b7ce272f745e2ae6178050ccb78fca515b13c0f9e92",
  "oldfee": 0.00000260,
  "fee": 0.00000520
}
```

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,267 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+                            "This command will fail if fee is not high enough or output is not large enough.\n"
+                            "User can specify totalFee, or use RPC setpaytxfee to set a higher fee rate.\n"
+                            "\nArguments:\n"
+                            "1. \"transactionid\"     (string, required) The txid to be bumped\n"
```
nit: Maybe use txid to correspond with the call header at L2587 or use `transactionid` there?

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,267 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+
+    if (request.fHelp || request.params.size() < 2 || request.params.size() > 3)
+        throw runtime_error(
+                            "bumpfee \"txid\" \"output\" \n"
```
add third parameter `("options")`?

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,267 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+                            "3. options               (object, optional)\n"
+                            "   {\n"
+                            "     \"confTarget\"      (int, optional) Confirmation target (in blocks)\n"
+                            "     \"totalFee\"        (numeric, optional) Total fee to pay, in satoshis (not btc)\n"
```
I'm not sure if we want absolute fee values here.
The user can't be sure how many inputs are getting added when setting this value, probably resulting in an uncontrollable feerate.

What about switching this to a feerate?

Victorsueca replied:
Another possibility is to leave the RPC with an absolute value and use Fee/KB on the GUI.

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,267 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    vector<COutput> vecOutputs;
+    pwalletMain->AvailableCoins(vecOutputs, false, NULL, true);
+    BOOST_FOREACH(const COutput& out, vecOutputs) {
+        if (out.tx->GetHash().GetHex() == hash.GetHex() && out.nDepth > 0)
```
This check looks really expensive for large wallets. Why not calling `mapWallet.find(hash)` and check `nDepth` (and maybe call `GetConflicts()` to ensure its not conflicted with a already mined tx)?

jonasschnelli commented:
Oh. I just realized that this PR does not add new inputs (it requires an output index to identify the change-output which then can be reduced).

IMO we should...

1.) not let the user identify which output is change
2.) allow bumping fees including adding new inputs (some transactions do not have a change output, some will not allow a reasonable bumping without adding a new input).

But 1.) & 2.) can also be solved later.

mrbandrews commented:
Feedback addressed.  Good catch on checking whether the tx had already been mined - that code was able to be shortened to a single line.
Yes, the approach of this PR is to solve 1) and 2) later.
If this looks good I'll squash the commits.

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,261 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    // This prevents us from trying to bump fee on a tx that has already been bumped.  So if the
+    // user created txid1, bumped the fee which resulted in txid2, and now wishes to bump the fee
+    // again, the user must call bumpfee on txid2.
+    const CWalletTx *conflictTx;
```
I don't think this section makes sense.  Transactions which are in newer in the wallet are not a good indication of the most "up to date" spend of the outputs.  If anything, conflicts in the mempool would be a good proxy, because at least for those you'd be able to calculate any descendants fees.  But it may make the most sense to make this more of a utility function and just assume the user is trying to bump the right txid.  I would just eliminate this set of checks entirely.

mrbandrews replied:
Fixed (removed)

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,261 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    }
+
+    // calculate the old fee and fee-rate
+    CAmount nDebit = wtx.GetDebit(ISMINE_SPENDABLE);
```
We need to be sure we are calculating the correct fee here.  Either using new `IsAllFromMe` or getting the fee from the mempool.

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,261 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        nNewFeeRate = CFeeRate(payTxFee.GetFeePerK());
+    }
+    else {
+        int estimateFoundTarget = newConfirmTarget;
```
you don't need this, you can just leave it out as an argument

mrbandrews replied:
Fixed.

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,261 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            // and run the command again.
+            //
+            CAmount nFeesWithDescendantsPlusRelay = it->GetModFeesWithDescendants() + ::minRelayTxFee.GetFee(txSize);
+            if (nNewFee < nFeesWithDescendantsPlusRelay) {
```
I think this reduces bumpfee to a two step process in the common case (whenever there are children) and makes it so the second step is going to require setting totalfee.  It seems more user friendly to just pay what it takes?

On a separate note, I think it might be nice to check all of the children transactions and make sure their ancestor fee rate isn't higher than the new feerate you are bumping to.  If one of them has a higher ancestor fee rate you are actually make your situation worse.

morcos replied:
I spoke with @sdaftuar and @mrbandrews about this and withdraw both of the above points.  If the transaction has children, it becomes quite complicated to decide whether bumping fee makes sense and by how much and its probably better to just report to the user the minimum fee that would be required to bump and let the user decide whether that makes sense.

mrbandrews commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,261 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+                            "User can specify totalFee, or use RPC setpaytxfee to set a higher fee rate.\n"
+                            "\nArguments:\n"
+                            "1. \"txid\"              (string, required) The txid to be bumped\n"
+                            "2. output                (int, required) The output to be decremented\n"
```
Done.

mrbandrews commented:
morcos feedback addressed, including adding the first commit from #9167.

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    const CWalletTx& wtx = pwalletMain->mapWallet[hash];
+
+    // check that we are not trying to bump a tx that has already been mined
+    if (wtx.GetDepthInMainChain() > 0)
```
You should probably check that the depth is exactly 0, because otherwise the tx is already mined or conflicted with a mined tx and no point in bumping.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+
+    // check that original tx signals opt-in-RBF
+    if (!SignalsOptInRBF(wtx))
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Transaction is not bip-125 replaceable");
```
"BIP 125"

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+
+    if (request.fHelp || request.params.size() < 2 || request.params.size() > 3)
+        throw runtime_error(
+                            "bumpfee \"txid\" output ( options ) \n"
```
Maybe move output into options? Even if it's required in this version, it should ideally become optional in the future...

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+
+    // calculate the old fee and fee-rate
+    CAmount nOldFee = wtx.GetDebit(ISMINE_SPENDABLE) - wtx.GetValueOut();
+    int txSize = (int)GetVirtualTransactionSize((CTransaction)wtx);
```
Just use `int64_t`? `int` is only guaranteed to be 15-bit, so it's too small anyway.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    CAmount nNewFee = 0;
+    CFeeRate nNewFeeRate;
+    if (payTxFee.GetFeePerK() > 0) {
+        nNewFeeRate = CFeeRate(payTxFee.GetFeePerK());
```
`nNewFeeRate = payTxFee;`?

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+
+    // if user set totalFee, use that instead
+    if (totalFee > 0) {
+        CAmount minTotalFee = nOldFeeRate.GetFee(txSize) + minRelayTxFee.GetFee(txSize);
```
`+ wtx.vin.size()` seems appropriate here too?

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+                std::string strError = strprintf("Insufficent fee due to the child transactions.\n");
+                strError += strprintf("The bumped fee must be at least: %s.\n", FormatMoney(nFeesWithDescendantsPlusRelay));
+                strError += strprintf("Number of child transactions: %u, total size of child transactions: %u\n", numDescendants, sizeDescendants);
+                strError += strprintf("Note that the child transactions would be evicted from the mempool and would not be mined.\n");
```
s/would/may probably/

In the future, it may make sense to combine some children transactions (ie, our own) into this one while bumping the fee.

luke-jr replied:
Also no strprintf here

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        poutput->nValue = poutput->nValue - nDelta;
+    }
+    else {
+        throw JSONRPCError(RPC_MISC_ERROR, "Output does not have enough money to bump the fee");
```
Why not just discard the change?

luke-jr reviewed (changes requested):
IMO the main logic should probably be split up into non-RPC files so it can be reused in the GUI...

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    // commit/broadcast the tx
+    CReserveKey reservekey(pwalletMain);
+    CWalletTx wtxBumped(pwalletMain, tx);
+    if (!pwalletMain->CommitTransaction(wtxBumped, reservekey, g_connman.get()))
```
This needs to be rebased...

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            if (nNewFee < nFeesWithDescendantsPlusRelay) {
+                uint64_t numDescendants = it->GetCountWithDescendants()-1;
+                uint64_t sizeDescendants = it->GetSizeWithDescendants() - it->GetTxSize();
+                std::string strError = strprintf("Insufficent fee due to the child transactions.\n");
```
No need for strprintf here

jonasschnelli commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+                            "3. options               (object, optional)\n"
+                            "   {\n"
+                            "     \"confTarget\":       \"n\",          (numeric, optional) Confirmation target (in blocks)\n"
+                            "     \"totalFee\":         \"n\",          (numeric, optional) Total fee to pay, in satoshis (not btc)\n"
```
I think, because every else use fee-ratres per KB, we should make clear at this point that `totalFee` and `maxFee` are absolute fee-values (and not per KB).

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2575,6 +2577,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+                strError += strprintf("The bumped fee must be at least: %s.\n", FormatMoney(nFeesWithDescendantsPlusRelay));
+                strError += strprintf("Number of child transactions: %u, total size of child transactions: %u\n", numDescendants, sizeDescendants);
+                strError += strprintf("Note that the child transactions would be evicted from the mempool and would not be mined.\n");
+                strError += strprintf("To avoid mempool eviction, consider bumping fee on the child transactions (with fee to pay for the ancestors).\n");
```
del strprintf

mrbandrews commented:
OK, rebased and squashed, with edits addressing the most recent feedback in a separate commit.

I didn't separate code into non-rpc files because I'm not 100% sure which logic should be moved, and thought this decision could be made when using it from the GUI, as moving the code then should be easy enough.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,262 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+
+    // if user set totalFee, use that instead
+    if (totalFee > 0) {
+        CAmount minTotalFee = nOldFeeRate.GetFee(txSize+wtx.vin.size()) + minRelayTxFee.GetFee(txSize+wtx.vin.size());
```
You should calculate txSize + wtx.vin.size() once and keep it in a const variable for these...

```C++
const int64_t maxNewTxSize = txSize + wtx.vin.size();
```

gmaxwell commented:
Asking the user to identify change is unreasonable and dangerous.  Within our own wallet we should know which outputs are our own.  This also has a problem of creating a mess when the original version of the transaction it spent, but later sends may have spent the replacements output.

I would suggest that a minimal bump-fee would do this:

(1) Only be available on transactions where none of their outputs have been spent (even in mempool).
(2) Preserve all the user provided outputs, potentially changing the change amount, potentially adding more inputs if needed to get past dust or just sufficient fees. Obey the opt-in RBF minimum increment
(3) Mark its local output(s) as unspendable until confirmed.

1/3 are required so that we don't end up with a mess when the 'wrong' version of the transaction confirms and invalidates the others.

A somewhat more advanced -- and perhaps better to do instead-- would be a "bumpunconfirmed" which would not act on a single transaction but on all valid unconfirmed sends of the local wallet at once--  generating a single replacement transaction which conflicts each of the originals, pays a higher fee, and marks its own output as unspendable until confirmed.

I suggest this might be better instead because it  would not need to have the requirement that its local outputs not be unspent, since it would replace all those spends at once. The reduced number of change outputs plus the ability to drop some of the inputs (so long as it still conflicts, and still marks them as unspendable until this transaction confirms) means that it can reduce the total transaction sizes.

mrbandrews committed:
[RPC] bumpfee

This command allows a user to increase the fee on a wallet transaction T, creating a "bumper" transaction B.
T must signal that it is BIP-125 replaceable.
T's change output is decremented to pay the additional fee.  (B will not add inputs to T.)
T cannot have any descendant transactions.
Once B bumps T, neither T nor B's outputs can be spent until either T or (more likely) B is mined.

Includes code by @jonasschnelli and @ryanofsky

mrbandrews commented:
Addressed the gmaxwell feedback as follows: now it identifies the change output and guards against spending outputs until the bumped transaction (or perhaps the original transaction) is mined.

I reorganized this into two commits, the latter being the python test, which is updated.

sdaftuar committed:
[wallet] Add IsAllFromMe: true if all inputs are from wallet

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2584,233 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Transaction is not BIP 125 replaceable");
+
+    if (wtx.mapValue.count("replaced_by_txid"))
+      throw JSONRPCError(RPC_INVALID_REQUEST, strprintf("Cannot bump transaction %s which was already bumped by transaction %s", hash.ToString(), wtx.mapValue.at("replaced_by_txid")));
```
Any reason not to automatically just switch to the new txid here?

mrbandrews replied:
For now, I'd rather leave it as is, and let the user resubmit with the correct txid.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2584,233 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    // figure out which output was change
+    // if there was no change output or if there were multiple change outputs, fail
+    int nOutput = -1;
+    for (int i=0; i < (int) wtx.tx->vout.size(); i++) {
```
Just use a size_t here, not cast to int... And use ++i, not i++

But more importantly, better to use C++11 iterators in this case:

```C++
for (const CTxOut& txout : wtx.tv->vout) {
    if (pwalletMain->IsChange(txout)) {
        if (nOutput != -1) {
            throw JSONRPCError(RPC_MISC_ERROR, "Transaction has multiple change outputs");
        }
        nOutput = i;
    }
}
```

sdaftuar replied:
@luke-jr  That won't compile since`i` is undefined...  Is there a better way to loop here where the index is needed?  Also I'd suggest that this kind of style nit can be cleaned up in a later pull anyway; this one has been open for long enough.

mrbandrews replied:
Changed to a size_t

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2584,233 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+
+    // mark the original tx as bumped
+    {
+        CWalletDB walletdb(pwalletMain->strWalletFile, "r+");
```
I think we're trying to keep CWalletDB only used by CWallet. @pstratem ?

luke-jr replied:
Probably this needs to go through CWallet::AddToWallet

mrbandrews replied:
I moved this code into a new method on the wallet so the CWalletDB is only used from CWallet.  I looked at AddToWallet but I would need to have edited it a bit, plus it does some stuff I don't need so I thought a new but concise method was a better approach.

luke-jr commented:
I wonder if this ought to interact with abandontransaction in some way?

sdaftuar commented on src/rpc/client.cpp:
```diff
@@ -112,6 +112,8 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "getmempoolancestors", 1 },
     { "getmempooldescendants", 1 },
+    { "bumpfee", 1 },
+    { "bumpfee", 2 },
```
I think this line should be removed now that the output is no longer specified?

sdaftuar reviewed (changes requested):
Looks pretty good to me overall, though there's one bug here (after #8850) that needs to be fixed, plus some doc fixes to reflect the changes that have been made in this PR from when it was first opened.

Also, I think bumpfee.py should be added to rpc-tests.py.

mrbandrews commented:
Pushed a new commit addressing the recent feedback and rebased due to a conflict in rpc-tests.py.

ryanofsky commented on qa/rpc-tests/bumpfee.py:
```diff
@@ -0,0 +1,198 @@
+        self.sync_all()
+        assert_equal(self.nodes[0].getbalance(), Decimal('0.01'))
+
+        # create and bump an RBF transaction
```
This test is long and it's not clear right away whether the different parts of it depend on each other. If it could be broken up in some way, I think it would be more readable and easier to work with in the future. E.g.

```
self.test_simple_bumpfee_succeeds(a1)
self.test_nonrbf_bumpfee_fails(a1)
self.test_notmine_bumpfee_fails(a0)
...
```

ryanofsky reviewed (approved):
Looks good. Lots of nitpicky comments, nothing major.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2584,228 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    // optional parameters
+    int newConfirmTarget = nTxConfirmTarget;
+    CAmount totalFee = 0;
+    CAmount bumpfeeMaxTxFee = 0.01 * COIN;
```
Regardless, we don't want to use flops for CAmount.

(I know that in this case the compiler can do static analysis and figure out the exact result, but we should not promote bad practice; It is proved to break on 32 bit platforms when static code analysis does not catch the flop.)

mrbandrews commented:
Pushed a new commit addressing ryanofsky's feedback.  

I removed the maxFee option.  The rationale for this option was to have a more conservative tx fee limit for a bumpfee user, as the default maxTxFee (0.1 btc) is quite high and in particular, if bumping fee on a transaction with lots of descendants a user might end up paying a pretty high fee for the parent transaction alone (since the descendants would be removed).  Since the new approach is to not allow bumpfee to be used on a tx with descendants at all, this rationale mostly goes away and I think the maxFee option is not worth it. 

I moved the python test above 'nulldummy.py' in the pull-tester list for the purpose of avoiding a conflict.

mrbandrews commented:
I got a travis failure I think due to the default transaction version=2, which caused a failure in the python test.  I corrected the test, rebased, and force-pushed.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2584,212 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+                            "bumpfee \"txid\" ( options ) \n"
+                            "\nBumps the fee of an opt-in-RBF transaction T, replacing it with a new transaction B.\n"
+                            "An opt-in RBF transaction with the given txid must be in the wallet.\n"
+                            "The command will not add new inputs or alter existing inputs.\n"
```
I don't think we should promise this...?

luke-jr replied:
I don't think we should guarantee this. (sorry for reposting)

ryanofsky replied:
Thanks, I changed the wording to describe the current behavior but suggest it could be improved in the future.

sdaftuar reviewed (changes requested):
A couple nits that you can take or leave, and one minor bug that should be fixed.  Did some light testing (including review of the bumpfee.py test included here) and this otherwise looks good to me.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -11,9 +11,12 @@
 #include "rpc/server.h"
+#include "script/sign.h"
 #include "timedata.h"
+#include "ui_interface.h"
```
Thanks, removed.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2584,212 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+                            + HelpExampleCli("bumpfee", "<txid>")
+                            );
+
+    RPCTypeCheck(request.params, boost::assign::list_of(UniValue::VSTR));
```
Fixed.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2584,212 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        if (totalFee < minTotalFee)
+            throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid totalFee, must be at least oldFee + relayFee: %s", FormatMoney(minTotalFee)));
+        nNewFee = totalFee;
+        nNewFeeRate = CFeeRate(totalFee, txSize);
```
Added a comment. I think this was probably intentional since it makes the fee rate check below more conservative.

ryanofsky replied:
Now fixed (got this backward).

sdaftuar commented:
ACK b1a06207ed57425e0e4cccaad4588485d4483b08 (feel free to squash).

I believe all of @gmaxwell's comments in https://github.com/bitcoin/bitcoin/pull/8456#issuecomment-264734557 have been addressed, with the exception of being able to add new inputs to a transaction to bump the fee.  I think that would be a fine improvement for the future, but I think this is still a useful feature even without it, so I don't think that should be a merge blocker.  @gmaxwell -- what do you think?

ryanofsky commented:
Squashed b1a06207ed57425e0e4cccaad4588485d4483b08 -> f4fa93b29a59c6d0f0341f442fd1805972b43ca1.

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2583,209 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        SignatureData sigdata;
+        if (!ProduceSignature(TransactionSignatureCreator(pwalletMain, &txNewConst, nIn, SIGHASH_ALL), scriptPubKey, sigdata))
+            throw JSONRPCError(RPC_WALLET_ERROR, "Can't sign transaction.");
+        input.scriptSig = sigdata.scriptSig;
```
@sdaftuar points out that this won't correctly update scriptWitness for witness transactions
Probably better to use the UpdateTransaction function

ryanofsky replied:
Added UpdateTransaction in e4a766291a90440d0ee8b8b25619cdd91329ffc2, and after a lot of struggle, eventually got a test to work which creates a segwit transaction and makes sure scriptWitness is set.

The test also uncovered another bug in the TransactionSignatureCreator call above where the SIGHASH_ALL value was incorrectly being interpretted as a CAmount. This is also fixed in e4a766291a90440d0ee8b8b25619cdd91329ffc2.

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2583,209 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            "If the change output is not big enough to cover the increased fee, the command will currently fail\n"
+            "instead of adding new inputs to compensate. (A future implementation could improve this.)\n"
+            "The command will fail if the wallet or mempool contains a transaction that spends one of T's outputs.\n"
+            "By default, the new fee will be calculated automatically using estimatefee/fallbackfee.\n"
```
nit: mentioning fallbackfee is too much implementation detail here

ryanofsky replied:
Removed in e4a766291a90440d0ee8b8b25619cdd91329ffc2.

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2581,6 +2583,209 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            if (newConfirmTarget <= 0) // upper-bound will be checked by estimatefee/smartfee
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid confTarget (cannot be <= 0)");
+        }
+        if (options.exists("totalFee")) {
```
Should there be a check against providing both options

ryanofsky replied:
Added in e4a766291a90440d0ee8b8b25619cdd91329ffc2.

morcos commented on qa/rpc-tests/bumpfee.py:
```diff
@@ -0,0 +1,223 @@
+    # of the mempool, then invalidate the block so the rbf transaction will be
+    # put back in the mempool. this makes it possible to check whether the rbf
+    # transaction outputs are spendable before the rbf tx is confirmed.
+    block = submit_block_with_tx(rbf_node, rbftx)
```
i'd abandon bumpid here so it doesn't accidentally rereplace rbfid if its reaccepted from the wallet

if later abandoning conflicts/replacements frees up the original to spend again then thats a change in behavior that can result in changing the test

ryanofsky replied:
Added abandontransaction call in b0d4f9323250d81b53668362413134bc7a13d998.

morcos commented:
ACK modulo comments above.

jtimon commented:
Concept ACK, needs rebase.

ryanofsky commented:
Rebased to handle named arguments.

morcos commented on qa/rpc-tests/bumpfee.py:
```diff
@@ -213,6 +213,10 @@ def test_unconfirmed_not_spendable(rbf_node, rbf_node_address):
 
+    # call abandontransaction to make sure the wallet does not add the bump
+    # transaction back into the mempool.
+    rbf_node.abandontransaction(bumpid)
```
This should take place after `submit_block_with_tx` and before `invalidateblock`
That is the only time it is guaranteed that bumpid is not in the mempool.

ryanofsky replied:
Oops, yes, I got confused about what the test was doing with the two transactions. Fixed and updated the comment in a1046e099e4fac8b82d6e268eeb9fb97e4a9c9ef.

gmaxwell commented:
I believe this should make the new transaction non-RBF replaceable.  (1) we currently won't replace it, (2) if someone made an original payment and the receiver is squaking because it's replaceable, a mechanism is needed to issue a non-replacable version.   Since we won't replace it, we can do this in one step.   Does this make sense?

gmaxwell commented:
Hm. I expected this to change the walletrbf default.

gmaxwell commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,211 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    // retrieve the original tx from the wallet
+    assert(pwalletMain != NULL);
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+    if (!pwalletMain->mapWallet.count(hash))
```
I can't see where it checks if the wallet is unlocked. Am I missing it?

morcos replied:
yikes that seems like a pretty big oversight

ryanofsky replied:
Added EnsureWalletIsUnlocked call in 59e425aa8e634a939e12594ce7c8824e9764e61c

laanwj commented:
> Hm. I expected this to change the walletrbf default.

I don't think we should change that last-minute for 0.14 (only mentioning that because this is tagged 0.14).

ryanofsky commented:
Created https://github.com/bitcoin/bitcoin/pull/9527 for changing the walletrbf default.

ryanofsky commented:
Added replaceable option in 8d3cd2844b844cad92d741d848ee0ffebdffef40, to allow users to create replacement transactions that are not replaceable.

I ran into a lot of spurious errors related to `fundrawtransaction` in the bumpfee tests while trying to add a simple test for this. fundrawtransaction would sometimes chose different sized inputs than the `small_output_fails` and `dust_to_fee tests` were expecting, so I rewrote these tests in 13960f870c1ba94f391012f9c9243c59951ea6ab to manually create RBF transactions with the right inputs.

I also ran into problems with `fundrawtransaction` using up all the right-sized inputs before these tests could use them, so I changed the order of the tests so they would run before any fundrawtransaction calls.

I also ran into problems with `fundrawtransaction` sometimes chosing a higher fee for the RBF transaction than the `test_rebumping` test expected, causing the bumpfee command in the test to fail with an "Invalid totalFee, must be at least ..." error. I did not figure out what was causing this, but it seemed to somehow be related to the number of `peer_node.sendtoaddress(rbf_node_address, 0.001)` calls made during the test setup.

sdaftuar commented:
re-ACK 8d3cd28

morcos commented:
tested ACK 8d3cd28

ryanofsky commented:
Squashed 8d3cd2844b844cad92d741d848ee0ffebdffef40 -> 1d71fd0dab96430643b818ea993acea2aaa09256

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,232 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    // Mark new tx not replaceable, if requested.
+    if (!replaceable) {
+        for (auto& input : tx.vin) {
+            if (input.nSequence < 0xfffffffe) input.nSequence = 0xfffffffe;
```
just set it

morcos replied:
I guess I don't see how it makes a difference, but it seems to me that the whole point of this PR is to change the transaction to the least degree possible.  I don't know why someone would have chosen to set their sequence to 0xffffffff, but if they did, why should we change it for them here?

luke-jr replied:
Agreed with @morcos, leave this as-is.

instagibbs reviewed (approved):
tACK with random nits. If you were to make one change I think printing out the necessary flat fee on rejection(for whatever reason) would be a solid improvement.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,232 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            "By default, the new fee will be calculated automatically using estimatefee.\n"
+            "The user can specify a confirmation target for estimatefee.\n"
+            "Alternatively, the user can specify totalFee, or use RPC setpaytxfee to set a higher fee rate.\n"
+            "At a minimum, the new fee rate must be high enough to pay a new relay fee and to enter the node's mempool.\n"
```
Added in 7c9ad69f946f1db623d4243aa4a10fe57296cc4d

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,232 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    hash.SetHex(request.params[0].get_str());
+
+    // retrieve the original tx from the wallet
+    assert(pwalletMain != NULL);
```
Removed in 7c9ad69f946f1db623d4243aa4a10fe57296cc4d

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,232 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    // (no point in bumping fee if we know that the new tx won't be accepted to the mempool)
+    // This may occur if the user set TotalFee or paytxfee too low, if fallbackfee is too low, or, perhaps,
+    // in a rare situation where the mempool minimum fee increased significantly since the fee estimation just a
+    // moment earlier. In this case, we report an error to the user, who may use totalFee to make an adjustment.
```
Added in 7c9ad69f946f1db623d4243aa4a10fe57296cc4d

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,232 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            "1. txid                  (string, required) The txid to be bumped\n"
+            "2. options               (object, optional)\n"
+            "   {\n"
+            "     \"confTarget\"        (numeric, optional) Confirmation target (in blocks)\n"
```
Options style here is the same as the style used in fundrawtransaction (and it seems clearer without all the n's and extra punctuation).

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,232 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            "2. options               (object, optional)\n"
+            "   {\n"
+            "     \"confTarget\"        (numeric, optional) Confirmation target (in blocks)\n"
+            "     \"totalFee\"          (numeric, optional) Total fee (NOT feerate) to pay, in satoshis\n"
```
User chosen feerate is possible with settxfee, and could be added as another option here in the future. Absolute fee is a reasonable thing a user might want to set, probably determining it from the fee of the previous transaction.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,232 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            "                         be left unchanged from the original. If false, any input sequence numbers in the\n"
+            "                         original transaction that were less than 0xfffffffe will be increased to 0xfffffffe\n"
+            "                         so the new transaction will not be explicitly bip-125 replaceable (though it may\n"
+            "                         still be replacable in practice if it has unconfirmed ancestors which are\n"
```
Referring to nodes that don't follow BIP-125? Maybe a more general note about these nodes would be helpful, but it seems like an odd thing to mention only in this context.

luke-jr replied:
"(though it may still be replacable in practice)" seems sufficient.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -826,6 +833,29 @@ void CWallet::MarkDirty()
+    wtx.mapValue["replaced_by_txid"] = newHash.ToString();
+
+    CWalletDB walletdb(strWalletFile, "r+");
+    walletdb.WriteTx(wtx);
```
Hum, should we be ignoring the return value here? (We check it in some places, ignore it in others in wallet.cpp, but it seems wrong to ignore it).

ryanofsky replied:
Will add a check.

sdaftuar replied:
I think we can just clean this up in a subsequent PR?  I agree that the differing behaviors are not good, but perhaps this is a small enough issue to not hold this up...

ryanofsky replied:
Either way seems fine. Added a check and error print for now in f9c4007977a25d53cc5ad9459db51a831fb8ae89.

TheBlueMatt replied:
I only brought it up because I'd prefer new code to check it...it seems bad to not do so.

ryanofsky replied:
Agreed, check is added in f9c4007977a25d53cc5ad9459db51a831fb8ae89.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -826,6 +833,29 @@ void CWallet::MarkDirty()
+    // Ensure for now that we're not overwriting data
+    assert(wtx.mapValue.count("replaced_by_txid") == 0);
+
+    wtx.mapValue["replaced_by_txid"] = newHash.ToString();
```
I'm super not convinced that this is sufficient. See all the places where isAbandoned is checked - we probably need to check for replacement in most, if not all, of the same places.

eg probably want to check in RelayWalletTransaction and possibly isSpent?

ryanofsky replied:
This is being set specifically to change the behavior of AvailableCoins to avoid creating new transactions which spend outputs of rbf & bumpfee transactions, for reasons described in comments in AvailableCoins.

I'm actually not sure why we don't mark the transaction that is being replaced to be abandoned. But I don't think it should make any practical difference right now.

RelayWalletTransaction in the normal case should already skip this transaction because it conflicts with the bumpfee transaction and won't be in mempool.

IsSpent could perhaps be written to not consider inputs of a transaction that's been replaced to be spent, though in practice it doesn't make a difference because the bumpfee replacing transaction always spends the same inputs as the replaced transaction.

sdaftuar replied:
Fee bumped should not imply abandoned -- it's entirely possible the original version of the transaction will confirm.

Changing IsSpent as you suggest ought to have no visible effect (since bumpfee spends all the same inputs, currently), yet I think it would be incorrect to make your change -- if we add smarter fee bumping behavior in the future, and it becomes possible to drop an input from the original tx in a replacement transaction, we would still want to treat the original input as spent until the transaction that spends it becomes conflicted (or the user explicitly abandons it).

In RelayWalletTransaction, I think it's more debatable whether to attempt relay of the original version, but since it would only succeed if somehow the feebumped version dropped out of your mempool, I think it's reasonable for the wallet to attempt to relay unless the user explicitly calls abandontransaction on it.  We don't care which version gets mined, and if something is subtly wrong with the bumped transaction then we might as well attempt the original.  The effect on tx relay bandwidth should be negligible.

Also -- I believe this is no different behavior than if we have conflicting transactions in our wallet that originate via some means other than bumpfee.  For instance if someone sends funds to us and then fee bumps it themselves, we'll wind up with both transactions in our wallet, and we'll attempt to relay both, which I think is correct behavior.

morcos replied:
100% agree with sdaftuar.  it would be incorrect to auto abandon the bumpee.

TheBlueMatt replied:
Abandoned doesn't imply the original version of the transaction won't confirm, it just means "pretend this isnt here, unless it somehow gets confirmed"?

But, ok, fair point regarding IsSpent - I'm not convinced thats is definitely The Right Behavior to not allow a user to spend something they removed from their transaction, but this is the more conservative option, so have no problem with it for now.

re: RelayWalletTransaction: I think this absolutely should have a replacement check...if a transaction has been replaced it shouldn't make it back into our mempool without the other transaction also replacing it, sure, but belt-and-suspenders. Plus a user could, theoretically, restart with a higher min relay fee (or upgrade to a new version), and this could break.

TheBlueMatt replied:
I'll withdraw my request after IRC discussion. I've been convinced that in some use-cases a user might prefer us to accept the old version of a transaction which has been bumped, so no need to change that.

ryanofsky replied:
Long discussion in IRC about this beginning here: https://botbot.me/freenode/bitcoin-core-dev/msg/79332875/.

I think the conclusion is that it is preferable to keep the current code which attempts to add all replaced and replacing transactions to the mempool , because there are cases when there is a problem with the replacing tx (e.g. output too close to dust limit) that would cause it to get rejected from mempool, and where it would be preferable to relay the original replaced transaction.

Also, the risk of the wallet relaying both replaced and replaced transactions to peers is very low. It will not happen unless a CWallet::ResendWalletTransactions call somehow is made before the initial CWallet::ReacceptWalletTransactions call in CWallet::postInitProcess finishes.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -826,6 +833,29 @@ void CWallet::MarkDirty()
+    CWalletDB walletdb(strWalletFile, "r+");
+    walletdb.WriteTx(wtx);
+
+    NotifyTransactionChanged(this, originalHash, CT_UPDATED);
```
Do we need/want a MarkDirty here, too? I think we might not need it, but it seems strange to not have it here when we have it in all the other similar things (conflicted, abandon, etc).

ryanofsky replied:
Will look more, but I would want to find a specific case it is needed before adding MarkDirty here. I have another PR https://github.com/bitcoin/bitcoin/pull/9381 which eliminates MarkDirty from AddToWallet() and will get rid of this entire MarkReplaced() method, replacing it with an AddToWallet call.

sdaftuar replied:
I don't think we should treat this differently from any other unconfirmed, conflicting transaction entering the wallet, so I don't believe this is necessary (and it seems confusing to me to add calls like that in places where we think that balances shouldn't be changing).

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1958,6 +2011,35 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
+            // intending to replace A', but potentially resulting in a scenario
+            // where A, A', and D could all be accepted (instead of just B and
+            // D, or just A and A' like the user would want).
+            if (nDepth == 0 && pcoin->mapValue.count("replaced_by_txid"))
```
Obviously prefer to leave this in because belt-and-suspenders, but shouldn't this check be redundant with the !InMempool() check, two checks above?

ryanofsky replied:
It's not completely redundant (see https://github.com/mrbandrews/bitcoin/blob/1d71fd0dab96430643b818ea993acea2aaa09256/qa/rpc-tests/bumpfee.py#L251 for the test which triggers this condition), but this is checking for a really esoteric case. See description above for all the gory details, but in this can happen when the replaced transaction A temporarily becomes part of a block instead of the replacing transaction B, and then gets added back to the mempool when there's a reorg.

TheBlueMatt reviewed (commented):
Only reviewed parts of the wallet.cpp stuff just yet, but figured I'd post it now.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,232 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        throw runtime_error(
+            "bumpfee \"txid\" ( options ) \n"
+            "\nBumps the fee of an opt-in-RBF transaction T, replacing it with a new transaction B.\n"
+            "An opt-in RBF transaction with the given txid must be in the wallet.\n"
```
It says in the wallet, not the mempool...?

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,236 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    bool replaceable = true;
+    if (request.params.size() > 1) {
+        UniValue options = request.params[1];
+        if (options.size() > 2)
```
`>3` now... but I don't know that we need an explicit check here. Seems like asking for bugs.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,236 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "confTarget and totalFee options should not both be set. Please provide either a confirmation target for fee estimation or an explicit total fee for the transaction.");
+        } else if (options.exists("confTarget")) {
+            newConfirmTarget = options["confTarget"].get_int();
+            if (newConfirmTarget <= 0) // upper-bound will be checked by estimatefee/smartfee
```
Should be `1` now (as of #9239). Maybe make a const int in some header for this number...

morcos replied:
nothing bad happens if you call it with 1, i think it makes most sense to allow users to select 1 meaning as fast as possible, and then we just return the fastest estimate we're comfortable with

luke-jr replied:
Okay

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,236 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+            if (newConfirmTarget <= 0) // upper-bound will be checked by estimatefee/smartfee
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid confTarget (cannot be <= 0)");
+        } else if (options.exists("totalFee")) {
+            totalFee = options["totalFee"].get_int();
```
`get_int64`

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,236 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        }
+
+        if (options.exists("replaceable") && !options["replaceable"].get_bool()) {
+            replaceable = false;
```
Prefer to just do:

```C++
if (options.exists("replaceable")) {
    replaceable = options["replaceable"].get_bool();
}
```

So that if the default changes, this still works right.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,236 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    if (poutput->nValue <= poutput->GetDustThreshold(::minRelayTxFee)) {
+        LogPrint("rpc", "Bumping fee and discarding dust output\n");
+        nNewFee += poutput->nValue;
+        tx.vout.erase(tx.vout.begin() + nOutput);
```
So long as the user is explicitly setting a total fee, we should either fail when discarding dust, or at least document this behaviour in the RPC help.

ryanofsky replied:
Added note to totalFee documentation in 23c389ab8697bc0040cfd45a32d282906554a84a.

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,236 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        // along with an exception. It would be good to return information about
+        // wtxBumped to the caller even if marking the original transaction
+        // replaced does not succeed for some reason.
+        throw JSONRPCError(RPC_WALLET_ERROR, "Unable to mark the original transaction as replaced.");
```
Maybe just add `"errors"` to the response below?

luke-jr replied:
I would worry that a caller might ignore the error and try again, but I guess the harm risk of that is minimal with bumpfee.

luke-jr reviewed (commented):
Almost there... (if you could use braces more, that'd be nice, but not critical)

ryanofsky reviewed (commented):
Implemented suggestions from @luke-jr in 23c389ab8697bc0040cfd45a32d282906554a84a, adding braces to all if statements, and making most of the other requested changes.

TheBlueMatt commented:
I know other ways of calling getbalance are pretty broken (#8183), but I think we need to fix a bit of how bumpfee interacts with getbalance "*". In testing I got getbalance "*" to give me a negative number, which seems super shitty.

Additionally, we need documentation on how bumpfee interacts with listunspent. While I dont mind the change outputs from bumpfee not appearing in the results, this is likely to be surprising to users (I believe this only otherwise occurs if the transaction was abandoned), so needs documented.

morcos commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2609,7 +2609,10 @@ UniValue bumpfee(const JSONRPCRequest& request)
             "     \"confTarget\"        (numeric, optional) Confirmation target (in blocks)\n"
-            "     \"totalFee\"          (numeric, optional) Total fee (NOT feerate) to pay, in satoshis\n"
+            "     \"totalFee\"          (numeric, optional) Total fee (NOT feerate) to pay, in satoshis.\n"
+            "                         In rare cases, the actual fee paid might be slightly higher than the specified\n"
```
I'd vote against making this statement.
I think it should be a more general principle of our wallet that you might always lose unexpectedly something on the order of the dust threshold, rather than explicitly mention it every time we think it could happen.   But if I'm outvoted, so be it..

morcos commented:
@TheBlueMatt I don't see how this PR has any affect on any of the getbalance calls.  I think what you are seeing is pre-existing misbehavior for multiple spends of the same outputs being present in the wallet.   

I went through all the calls to AvailableCoins (which is the only thing this PR changes) and I think the only open decision is deciding what we want the interaction with listunspent to be and documenting it.

TheBlueMatt commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2584,258 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    } else {
+        // use the user-defined payTxFee if possible, otherwise use smartfee / fallbackfee
+        nNewFeeRate = payTxFee;
+        if (nNewFeeRate.GetFeePerK() == 0) {
```
I think if you have specified a manual confTarget you probably want to always take this branch.

ryanofsky replied:
Done in ceeecad175b7e9f8c373f396ecc610c9ecb9c4b8.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1958,6 +2017,37 @@ void CWallet::AvailableCoins(vector<COutput>& vCoins, bool fOnlyConfirmed, const
+            // be a 1-block reorg away from the chain where transactions A and C
+            // were accepted to another chain where B, B', and C were all
+            // accepted.
+            if (nDepth == 0 && pcoin->mapValue.count("replaces_txid")) {
```
I believe this and the check further down should have a fOnlyConfirmed && at the beginning to avoid not showing the tx currently in mempool's change output in listunspent, which I think would be the correct behavior.

TheBlueMatt replied:
Or maybe only the one below, but in either case should be well-documented.

ryanofsky replied:
Moved these checks into from `CWallet::AvailableCoins` to `CWalletTx::IsTrusted` in 32c01603a90a08459814852ba0c531c5b08091ec.

ryanofsky committed:
[wallet] Add include_unsafe argument to listunspent RPC

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2413,11 +2417,17 @@ UniValue listunspent(const JSONRPCRequest& request)
     assert(pwalletMain != NULL);
     LOCK2(cs_main, pwalletMain->cs_wallet);
-    pwalletMain->AvailableCoins(vecOutputs, false, NULL, true);
+    pwalletMain->AvailableCoins(vecOutputs, !include_untrusted, NULL, true);
```
note: in the future we should rename and invert the AvailableCoins argument to this.

ryanofsky commented:
Squashed 28fd457a73fc6291f72d89d9dd3532ada9ebe35c -> ce3a363049bf8fdd8be926ca3ab8c803c98de6d2.

TheBlueMatt commented:
utACK ce3a363049bf8fdd8be926ca3ab8c803c98de6d2

morcos commented:
tested previous iteration ACK ce3a363

luke-jr commented on src/rpc/server.cpp:
```diff
@@ -79,16 +79,20 @@ void RPCTypeCheck(const UniValue& params,
     }
 }
 
+void RPCTypeCheckArgument(const UniValue& value, UniValue::VType typeExpected)
```
This seems redundant. Type-checking is the purpose of the .get_&lt;type> methods on UniValue.

TheBlueMatt replied:
We absolutely do not use .get_<type>'s built-in type check anywhere - it throws a different type of exception which will not be correctly reported as an RPC_TYPE_ERROR.

sdaftuar commented:
utACK ce3a363049bf8fdd8be926ca3ab8c803c98de6d2 (reviewed diff from last iteration I tested and ACKed above).

luke-jr reviewed (approved):
Looks okay. A few suggested changes, but not a big deal if they don't go in.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2364,6 +2366,10 @@ UniValue listunspent(const JSONRPCRequest& request)
+            "4. include_unsafe (bool, optional, default=true) Include outputs that are not safe to spend\n"
+            "                  because they come from unconfirmed untrusted transactions or unconfirmed\n"
+            "                  replacement transactions (cases where we can't be sure a conflicting\n"
+            "                  transaction won't be mined).\n"
```
Changed wording in 4beb7c92c70b7668042383d37c49545198aa00cd.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2350,9 +2352,9 @@ UniValue listunspent(const JSONRPCRequest& request)
+    if (request.fHelp || request.params.size() > 4)
         throw runtime_error(
-            "listunspent ( minconf maxconf  [\"addresses\",...] )\n"
+            "listunspent ( minconf maxconf  [\"addresses\",...] [include_unsafe] )\n"
```
Interesting. I wasn't following the named argument discussion in detail, but I thought named arguments would be preferable to options because they are easier to specify on the command line. But maybe there are advantages to options objects that I'm not aware of.

ryanofsky commented:
Squashed 4beb7c92c70b7668042383d37c49545198aa00cd -> cc0243ad32cee1cc9faab317364b889beaf07647

morcos commented:
Still ACK cc0243a

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2595,6 +2597,261 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+    }
+
+    if (wtx.mapValue.count("replaced_by_txid")) {
+        throw JSONRPCError(RPC_INVALID_REQUEST, strprintf("Cannot bump transaction %s which was already bumped by transaction %s", hash.ToString(), wtx.mapValue.at("replaced_by_txid")));
```
ultra-nit: "bumped" is a bit confusing since it's already used in another sense(bumped fee), "replace" is likely better.

instagibbs reviewed (approved):
tACK cc0243ad32cee1cc9faab317364b889beaf07647

no need to change nit pre-merge, very minor

jameshilliard commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2582,6 +2597,261 @@ UniValue fundrawtransaction(const JSONRPCRequest& request)
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Transaction has been mined, or is conflicted with a mined transaction");
+    }
+
+    if (!SignalsOptInRBF(wtx)) {
```
Should probably have an option to override this.

ryanofsky replied:
An option to signal RBF in the original transaction (which already exists, see -walletrbf option, and #9592, #9527), or an option to create a bumpfee transaction even if the original doesn't signal RBF?

jameshilliard replied:
An option to create a bumpfee transaction regardless of if the original signaled RBF since some nodes may accept it.

ryanofsky replied:
There was some discussion about this idea in IRC https://botbot.me/freenode/bitcoin-core-dev/msg/79836432/ (unclear what the outcome was)
